crypto/src/lib.rs

```diff
@@ -278,79 +308,167 @@ impl From<KeyPair> for (PublicKey, PrivateKey) {
 
 ffi::ffi_item! {
     /// Public Key used in signatures.
-    #[derive(Clone, PartialEq, Eq, PartialOrd, Ord, CopyGetters)]
-    #[cfg_attr(not(feature="ffi_import"), derive(DebugCustom, Display, Hash, DeserializeFromStr, SerializeDisplay, Decode, Encode, IntoSchema))]
-    #[cfg_attr(not(feature="ffi_import"), debug(fmt = "{{digest: {digest_function}, payload: {}}}", "self.normalize()"))]
-    #[cfg_attr(not(feature="ffi_import"), display(fmt = "{}", "self.normalize()"))]
-    pub struct PublicKey {
-        /// Digest function
-        #[getset(get_copy = "pub")]
-        digest_function: Algorithm,
-        /// Key payload
-        payload: ConstVec<u8>,
+    #[derive(Clone, PartialEq, Eq)]
+    #[cfg_attr(not(feature="ffi_import"), derive(DeserializeFromStr, SerializeDisplay))]
+    #[cfg_attr(all(feature = "ffi_export", not(feature = "ffi_import")), ffi_type(opaque))]
+    #[allow(missing_docs)]
+    pub enum PublicKey {
+        Ed25519(ed25519::PublicKey),
+        Secp256k1(secp256k1::PublicKey),
+        BlsNormal(bls::BlsNormalPublicKey),
+        BlsSmall(Box<bls::BlsSmallPublicKey>),
     }
 }
```

Comment:
```suggestion
    pub enum PublicKey<A: Algorithm>(payload: A::Payload)
}
```

I'm just wondering something. Can we make `PublicKey` generic like this?

Reply:
the first thing I notice is that this way in `p2p` crate you would see the benefit immediately because there we use a specific public key type. Idk, if there are other places to benefit from this

Reply:
I think only p2p uses a fixed algorithm, while account keys, for example, can be whatever the user wants/needs and this wouldn't be represented all too well.

It would be nice to make smth that would work better with the "we know the type of the key here" for p2p, but other code that is actually dynamic over the algorithm shouldn't struggle with it.

We _may_ consider exposing the individual signature algorithm and the types of their keys, but this is quite a considerable increase of API surface just for the p2p.

Maybe we can just use whatever crates we need in p2p directly? It's not like `iroha_crypto` is bringing a lot of value to p2p in this case, there's nothing to abstract over (unlike `data_model`, which needs to support different signature algos)

Reply:
I partly agree with @DCNick3 that looks like there are no benefits rather than p2p. The whole point of our key types is that their API is algorithm-agnostic so that it would be easier to introduce new crypto algorithms in future and etc.

However I'd prefer to see all crypto code in just `iroha_crypto` crate. That's easier to audit and more incapsulated.

We might also introduce the support of other crypto key types in p2p as well